sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeArrayWriter.java

```diff
@@ -33,91 +38,144 @@
   // The offset of the global buffer where we start to write this array.
   private int startingOffset;
 
+  // The number of elements in this array
+  private int numElements;
+
+  private int headerInBytes;
+
+  private void assertIndexIsValid(int index) {
+    assert index >= 0 : "index (" + index + ") should >= 0";
+    assert index < numElements : "index (" + index + ") should < " + numElements;
+  }
+
   public void initialize(BufferHolder holder, int numElements, int fixedElementSize) {
-    // We need 4 bytes to store numElements and 4 bytes each element to store offset.
-    final int fixedSize = 4 + 4 * numElements;
+    // We need 4 bytes to store numElements in header
+    this.numElements = numElements;
+    this.headerInBytes = calculateHeaderPortionInBytes(numElements);
 
     this.holder = holder;
     this.startingOffset = holder.cursor;
 
-    holder.grow(fixedSize);
-    Platform.putInt(holder.buffer, holder.cursor, numElements);
-    holder.cursor += fixedSize;
+    // Grows the global buffer ahead for header and fixed size data.
+    holder.grow(headerInBytes + fixedElementSize * numElements);
+
+    // Initialize information in header
```

Comment:
we should explain it more clear, e.g. write `numElements` to the head and clear out the null bits